### PR TITLE
add support for async galleries

### LINKED ISSUE #88: Support for async examples?
When I'm inside a `jupyter` notebook, the following is a valid code block:

```python
import asyncio
import time

start = time.time()
await asyncio.sleep(1)
stop = time.time()
f"I waited for {stop - start} seconds!"
```

However, if you put this inside a gallery script and try to run this, you'll hit 

```
SyntaxError: 'await' outside function
```

This is expected, since we are currently not inside an `async` context. The difference to `jupyter` notebooks is that they have an `async` event loop in the background.

I would like to have the same convenience for example scripts as well. Without, it is a pain to write example that have `async` code in it. Plus it will become nightmarish if one at the same time also wants to support the downloadable jupyter notebook.

If this is something that we want, I'm happy to help designing and implementing this feature.

### PR DESCRIPTION
**Pull request by pmeier**
This Pull request fixes #88.

 - [x] I completed the issue numbers (#xx) in the sentence above. The word "fixes" should remain in front of each issue
 - [x] My PR is tagged as draft when I'm still working on it, and I remove the draft flag when it is ready for review.
- [x] I added one or several `changelog.md` entries in the appropriate "in progress" section (not the last release one)

### b - My PR adds some features:

*(Delete this section if not relevant)*

 - [x] Each issue is well-described, well-labeled and explains/agrees on the expected solution (high-level).
 - [x] I added documentation gallery examples showing the new features
 - [x] I created tests for each new feature. In particular I have for each feature:
   - [x] nominal tests (several simple ones or a parametrized one, or both)
   - [x] edge tests (valid but very particular cases, for example empty data, daylight savings change day, sparse matrix, etc.)
   - [x] error tests (in which case use `with pytest.raises(MyErrorType, match="...")`)

### CODE REVIEW

**Comment by pmeier** [comment]
This already works somewhat:

image

A few issues though that I need to think about how to solve. Since we wrap the whole block in an async function, 

1. we need to return the result for it to be displayed, and
2. we need to somehow propagate the locals inside the wrapper function into the globals of the overall script.

**Comment by pmeier** [comment]
Ok, maybe the issues in https://github.com/smarie/mkdocs-gallery/pull/90#issuecomment-1858627875 aren't too bad after all:

1. We can use something similar to 
   https://github.com/smarie/mkdocs-gallery/blob/cc5c4ad508c03df38c3f58c2c0b8a956d458b900/src/mkdocs_gallery/gen_single.py#L644-L649

   to capture the output that we want to return. The fact that this happens on the AST here makes me believe that we want to perform the patching on the AST level as well.

2. Before returning, we can just simply update `script.run_vars.fake_main`

   https://github.com/smarie/mkdocs-gallery/blob/cc5c4ad508c03df38c3f58c2c0b8a956d458b900/src/mkdocs_gallery/gen_single.py#L652

   with the locals in the function.

File: src/mkdocs_gallery/gen_single.py [hunk 1]
```diff
@@ -739,6 +739,66 @@ def _reset_cwd_syspath(cwd, path_to_remove):
...
+
+
+def _needs_async_handling(bcontent, *, compiler_flags) -> bool:
+    try:
+        compile(bcontent, "<_needs_async_handling()>", "exec", compiler_flags, dont_inherit=1)
+    except SyntaxError as error:
+        # FIXME
+        #  bool(re.match(r"'(await|async for|async with)' outside( async)? function", str(error)))
+        #  asynchronous comprehension outside of an asynchronous function
+        return "async" in str(error)
```
Comment (pmeier) [inline src/mkdocs_gallery/gen_single.py:756]: As far as I'm aware there are four error messages that we need to look out for:
- await outside function
- async for outside async function
- async with outside async function
- asynchronous comprehension outside of an asynchronous function

Kind of a bummer that they don't follow the same pattern, but here we are. We can still fit this into a reasonable regex.

A simpler, but less robust solution (and yes I'm aware that using the error message as trigger is not robust in the first place), would be to just check for `"async"` as implemented right now. LMK what you think.

File: src/mkdocs_gallery/gen_single.py [hunk 2]
```diff
@@ -739,6 +739,66 @@ def _reset_cwd_syspath(cwd, path_to_remove):
...
+    if isinstance(last_node, ast.Expr):
+        # FIXME
+        #  ast.Assign(targets=[ast.Name(id="__async_wrapper_result__", ctx=ast.Store())], value=last_node)
+        last_node = compile(
+            f"__async_wrapper_result__ = ({ast.unparse(last_node)})",
+            "<_apply_async_handling()>",
+            "exec",
+            compiler_flags | ast.PyCF_ONLY_AST,
+            dont_inherit=1,
+        ).body[0]
```
Comment (pmeier) [inline src/mkdocs_gallery/gen_single.py:793]: For some reason, `ast.Assign`ing the `ast.Expr`ession here does not work. With the code that I nicked from

https://github.com/smarie/mkdocs-gallery/blob/cc5c4ad508c03df38c3f58c2c0b8a956d458b900/src/mkdocs_gallery/gen_single.py#L654-L655

I always end up with something like

```pycon
>>> print(ast.unparse(ast.fix_missing_locations(foo)))
__async_wrapper_result__ = 
f'I waited for {stop - start} seconds!'
```

Note the missing parentheses that would allow this as multiline statement. I need to look into why this is happening and fix it. However, for the first iteration, just unparsing the last node, create the assignment as code, and parse it again, works well enough.

**Review by pmeier — COMMENTED** [review]
@smarie Could you make a first pass on this? I deviated a little from our plan in #88. Instead of having a feature flag for the async handling, I've opted to detect it, by checking if compiling the code leaves us with an async error. IMO this is robust enough, but I'll let you be the judge of that. The patch will also work with a feature flag, so we can easily switch to that if you prefer.

**Comment by pmeier** [comment]
Don't worry about the 3.7 and 3.8 CI jobs that are failing. [`ast.unparse`](https://docs.python.org/3/library/ast.html#ast.unparse) is only available starting `3.9`. And that will ultimately removed as explained in https://github.com/smarie/mkdocs-gallery/pull/90#discussion_r1429316074.

File: src/mkdocs_gallery/gen_single.py [hunk 3]
```diff
@@ -788,9 +848,7 @@ def execute_code_block(compiler, block, script: GalleryScript):
 
     try:
         ast_Module = _ast_module()
-        code_ast = ast_Module([bcontent])
```
Comment (smarie) [inline src/mkdocs_gallery/gen_single.py:791]: I was surprised to see this but it is indeed present in sphinx-gallery: https://github.com/sphinx-gallery/sphinx-gallery/blob/master/sphinx_gallery/gen_rst.py#L965

They probably use this as a pre-validator of good syntax. Should we keep it too ? In that case I suggest to keep these original lines and simply add your `if _needs_async_handling` extra two lines below (dropping `_parse_code`). That way the change will be easier to spot when comparing with sphinx-gallery code

File: src/mkdocs_gallery/gen_single.py (hunk 1 above)
Comment (smarie) [reply src/mkdocs_gallery/gen_single.py:756]: From what I understand from https://github.com/ipython/ipython/pull/10390/files#diff-6abb45111ba17c2eb57fa3302b2e73aec68a5adfa6805dcf90d3fe1e9e72bd69R61

Their solution is even simpler: they try to apply the async patch, and if it works then it means that it was needed :)

Could the same work for us ?

Comment (pmeier) [reply src/mkdocs_gallery/gen_single.py:756]: Sure, we can do that as well. Let me fix it.

File: src/mkdocs_gallery/gen_single.py (hunk 3 above)
Comment (pmeier) [reply src/mkdocs_gallery/gen_single.py:791]: > They probably use this as a pre-validator of good syntax. Should we keep it too ?

That is what we basically have right now. If we find a `SyntaxError` there, we need to check whether this comes from using async stuff and can't just exit the `try` branch pre-maturely. In the light of https://github.com/smarie/mkdocs-gallery/pull/90#discussion_r1434048801, we could just bubble up a `SyntaxError` if our async handling doesn't resolve it.

File: src/mkdocs_gallery/gen_single.py (hunk 1 above)
Comment (pmeier) [reply src/mkdocs_gallery/gen_single.py]: Implemented in 5e0e63693b5c951687d7f8519ec253a84d200ca8.

File: src/mkdocs_gallery/gen_single.py (hunk 3 above)
Comment (pmeier) [reply src/mkdocs_gallery/gen_single.py:791]: I've implemented my suggested in 5e0e63693b5c951687d7f8519ec253a84d200ca8. Please have a look.

File: src/mkdocs_gallery/gen_single.py (hunk 2 above)
Comment (pmeier) [reply src/mkdocs_gallery/gen_single.py]: Fixed in 031bacdddaeb0a3a85c6f4fa8f089fe5f35a6cf7.

**Comment by pmeier** [comment]
@smarie I'm happy with the functionality. Other than https://github.com/smarie/mkdocs-gallery/pull/90#discussion_r1434036881 do you have any other comments that could prevent this from being merged? If no, I'll get on improving the example and documenting this change in general.

As for tests, I see that a lot of them are commented out. How do you want to handle this? Just trust the fact that we can build the examples including the async one?

**Comment by smarie** [comment]
Thanks @pmeier ! I'll have a more thorough look in the upcoming days, just a quick answer concerning the tests: this project was a port of sphinx-gallery "as close as possible to the original", so that evolutions/improvements/bugfixes could easily be performed on both sides. Obviously this was not ideal (creating a common core lib would do much better), and that forced me to stick with some coding style that I would not have performed from scratch. A side effect of trying to not change much the code design, was that I was totally unable to integrate the original tests - too far from what I would have intuitively done, and obviously very sticky with the design of the internals.

Lacking time, I decided to go for a very crude strategy: the doc page of this project IS the main test. So if you add a gallery example for async, it will be taken into account for coverage for example. That being said, adding a few low-level tests dedicated to your utility functions could make this much more robust to edge cases, and error cases.

**Comment by pmeier** [comment]
@smarie I've added some text to the example as well as tests. This is ready for another round of review.

**Comment by pmeier** [comment]
Ping @smarie

**Comment by pmeier** [comment]
Ping @smarie

File: docs/examples/plot_12_async.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
+"""
+# Support for asynchronous code
```
Comment (smarie) [inline docs/examples/plot_12_async.py:2]: Very nice example @pmeier ! I have two comments on the text :

 - The text could maybe be improved in order to be more oriented towards the feature that is now available, and not towards the path that led to this feature being available. I'll try to suggest some modifications in this sense.
 - if all of this is automatic without particular configuration we should maybe say it explicitly. Again, I'll try to suggest something.

File: docs/examples/plot_12_async.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+downside is that it "poisons" your code base. If you want to `await` a coroutine, you have to be inside a `async def`
+context. Doing so turns the function into a coroutine function and thus forces the caller to also `await` its results.
+Rinse and repeat until you reach the beginning of the stack.
+
+While this might be acceptable or not even an issue for applications, e.g. a web frameworks, for scripts it is usually
+a nuisance. [`jupyter` notebooks](https://jupyter.org/), or rather the [`IPython` kernel](https://ipython.org/) running
+the code inside of them, have some
+[background handling to allow top-level asynchronous code](https://ipython.readthedocs.io/en/stable/interactive/autoawait.html).
+
+And so does `mkdocs-gallery` to keep examples that require asynchronous code nice and clean.
```
Comment (smarie) [inline docs/examples/plot_12_async.py:16]: This is only a suggestion, feel free to reject it - I'll merge whatever your decision.

```suggestion
[PEP 429](https://peps.python.org/pep-0492), which was first implemented in
[Python 3.5](https://docs.python.org/3/whatsnew/3.5.html#whatsnew-pep-492), added initial syntax for asynchronous
programming in Python: `async` and `await`. 

While this was a major improvement in particular for UX development, one major
downside is that it "poisons" the caller's code base. If you want to `await` a coroutine, you have to be inside a `async def`
context. Doing so turns the function into a coroutine function and thus forces the caller to also `await` its results.
Rinse and repeat until you reach the beginning of the stack.

Since version `0.10.0`, `mkdocs-gallery` is now able to automatically detect code blocks using async programming, and to handle them nicely so that you don't have to wrap them. This feature is enabled by default and does not require any configuration option. This feature makes it consistent with [`jupyter` notebooks](https://jupyter.org/), or rather the [`IPython` kernel](https://ipython.org/) running
the code inside of them, that is equipped with 
[background handling to allow top-level asynchronous code](https://ipython.readthedocs.io/en/stable/interactive/autoawait.html).
```

Comment (pmeier) [reply docs/examples/plot_12_async.py:16]: Sounds good to put more focus on `mkdocs-gallery`. Thanks!

**Comment by smarie** [comment]
All set, thanks a lot for this nice PR @pmeier !
